packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -935,16 +935,21 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       parseConstituentType: () => N.TsType,
       operator: TokenType,
     ): N.TsType {
-      this.eat(operator);
+      const leadingOperator = this.eat(operator);
+      const leadingOperatorTokStartLoc = this.state.lastTokStartLoc;
+      const leadingOperatorTokStart = this.state.lastTokStart;
```

Comment:
```suggestion
      const node: N.TsUnionType | N.TsIntersectionType = this.startNode();
      const leadingOperator = this.eat(operator);
```
We can always start node earlier.

Reply:
there is one more case that should be taken into account:
```
type M = | number
```

typescript in this case generates union type to and babel does not